doc/kv2json.1
.Dd April 4, 2018
.Dt kv2json 1
.Os
.Sh NAME
.Nm kv2json
.Nd convert key-value input to json
.Sh SYNOPSIS
.Nm
.Op Fl _h
.Op Fl c Ar char
.Op Fl o Ar name
.Op Fl r Ar sep
.Op Fl s Ar sep
.Sh DESCRIPTION
The
.Nm
tool takes plain text key-value data as input and
converts it into json.
.Sh OPTIONS
The following options are supported by
.Nm :
.Bl -tag -width c_char_
.It Fl _
Replace spaces in keys with an underscore (_).
.It Fl c Ar char
Treat 'char' as a comment character or string and
ignore any input including and up to the end of the
line.
Defaults to '#'.
.It Fl h
Display help and exit.
.It Fl o Ar name
Specify the object name to use for non-key-value
prefixes for multiple records.
Defaults to "name".
.It Fl r Ar sep
Specify the record separator to differentiate multiple
elements.
Defaults to an empty line.
.It Fl s Ar sep
Specify the key-value separator.
Defaults to '='.
.El
.Sh DETAILS
Many unix tools and configuration files utilize
trivial key-value pairs as input; many other tools
prefer to use more structured JSON.
The
.Nm
utility can be used to trivially generate JSON from
the input by converting multiple sequences of
key-value separated items into JSON objects.
.Pp
Multiple records or objects can be identified by
separating them via an empty line or by using the
.Fl r
flag.
.Pp
Any non-key-value text immediately preceeding a
key-value block will be interpreted as the name of the
object.
See EXAMPLES for an illustration of this feature.
.Pp
.Nm
reads data from STDIN and writes output to STDOUT.
Whitespace at the beginning or end of the line as well
as around the key-value separator is ignored,
as are comments and multiple empty lines.
.Sh EXAMPLES
The following examples illustrate common usage of this tool.
.Pp
.Bd -literal -offset indent
cat <<EOF | kv2json
> key1 = value1
> key2 = value2
> EOF
[
   {
     "key1" : "value1",
     "key2" : "value2"
   }
]
.Ed
.Pp
To parse multiple records with an object identifier
("name"):
.Bd -literal -offset indent
cat <<EOF | kv2json -s :
> item1
> key1 : value1
> # ignore this line
> key2 : value2
>
> item2
> key1 : value1
> EOF
[
   {
      "key1" : "value1",
      "key2" : "value2",
      "name" : "item1"
   },
   {
      "key1" : "value1",
      "name" : "item2"
   }
]
.Ed
.Pp
Note that duplicate keys will yield an array.
Here, we use "---" as the record separator and "id" as
the object identifier:
.Bd -literal -offset indent
cat <<EOF | kv2json -o id -r ---
> some sort of identifier
> key1 = some value
> key1 = some other value
> key2 = val1
> ---
> second object
> key1 = some value
> key2 = val1
> key2 = val2
> EOF
[
   {
      "id" : "some sort of identifier",
      "key1" : [
         "some value",
         "some other value"
      ],
      "key2" : "val1"
   },
   {
      "id" : "second object",
      "key1" : "some value",
      "key2" : [
         "val1",
         "val2"
      ]
   }
]
.Ed
.Sh EXIT STATUS
.Ex -std 
.Sh SEE ALSO
.Xr jq 1
.Sh HISTORY
.Nm
was originally written by
.An Jan Schaumann
.Aq jschauma@netmeister.org
in March 2018.
.Sh BUGS
Please report bugs and feature requests to the author
via email.


CHANGES
Changelog:

Version 1.1 (2018-04-04)
* underscore whitespace in key names

Version 1.0 (2018-04-03)
* initial version of kv2json


src/kv2json.pl
#! /usr/local/bin/perl -Tw
#
# Copyright 2018 Oath, Inc.
#
# Originally written by Jan Schaumann <jschauma@netmeister.org> in March 2018.
#
# This tool converts key-value data into json output.
# See the manual page for details.
#
# Licensed under the terms of the BSD License.
# See LICENSE file in project root for terms.
#
# https://github.com/jschauma/kv2json

use 5.008;

use strict;
use File::Basename;
use Getopt::Long;
Getopt::Long::Configure("bundling");

use JSON;

###
### Constants
###

use constant TRUE => 1;
use constant FALSE => 0;

use constant EXIT_FAILURE => 1;
use constant EXIT_SUCCESS => 0;

###
### Globals
###

my $PROGNAME = basename($0);
my %OPTS = (
		"comment"   => "#",
		"object"    => "name",
		"record"    => "",
		"separator" => "=",
	   );

my @J;

###
### Subroutines
###

sub addItemToHash($$$) {
	my ($h, $k, $v) = @_;
	my %hash = %{$h};
	# Multiple values for the same key yield
	# a JSON array rather than overriding
	# previously found keys.
	if (exists($hash{$k})) {
		my @a;

		if (ref($hash{$k}) eq 'ARRAY') {
			@a = @{$hash{$k}};
		} else {
			push(@a, $hash{$k});
		}
		push(@a, $v);
		$hash{$k} = \@a;
	} else {
		$hash{$k} = $v;
	}

	return %hash;
}

sub init() {
	my ($ok);

	$ok = GetOptions(
			"underscore|_"  => \$OPTS{'_'},
			"comment|c=s" 	=> \$OPTS{'comment'},
			"help|h" 	=> \$OPTS{'help'},
			"object|o=s" 	=> \$OPTS{'object'},
			"record|r=s"	=> \$OPTS{'record'},
			"separator|s=s"	=> \$OPTS{'separator'},
			);

	if (scalar(@ARGV)) {
		print STDERR "I can't deal with spurious arguments after flags.  Try -h.\n";
		exit(EXIT_FAILURE);
		# NOTREACHED
	}

	if ($OPTS{'help'} || !$ok) {
		usage($ok);
		exit(!$ok);
		# NOTREACHED
	}
}

sub parseInput() {
	my $sepCounter = 0;
	my %hash;

	foreach my $line (<STDIN>) {
		$line =~ s/^\s*//;
		$line =~ s/\s*$//;

		my $c = $OPTS{'comment'};
		if ($line =~ m/$c/) {
			$line =~ s/$c.*//;
			if ($line =~ m/^$/) {
				next;
			}
		}

		my $r = $OPTS{'record'};
		if ($line =~ m/^$r$/) {
			# Skip over multiple separators only if it's
			# the default to avoid the creation of empty
			# objects.
			if (($r eq "") && ($sepCounter > 0)) {
				next;
			}
			$sepCounter++;
			if (scalar(keys(%hash)) > 0) {
				my %h = %hash;
				push(@J, \%h);
			}
			%hash = ();
			next;
		} else {
			$sepCounter = 0;
		}

		my $s = $OPTS{'separator'};
		if ($line =~ m/$s/) {
			my @kv = split(/$s/, $line, 2);
			my $k = $kv[0];
			my $v = $kv[1];

			$k =~ s/^\s*//;
			$v =~ s/^\s*//;
			$k =~ s/\s*$//;
			$v =~ s/\s*$//;

			if ($OPTS{'_'}) {
				$k =~ s/ /_/g;
			}

			%hash = addItemToHash(\%hash, $k, $v);
		} else {
			%hash = addItemToHash(\%hash, $OPTS{'object'}, $line);
		}
	}

	# Last entry may not have a separator.
	if (scalar(keys(%hash)) > 0) {
		push(@J, \%hash);
	}
}

sub printOutput() {
	if (scalar(@J) > 0) {
		my $json = JSON->new->allow_nonref->canonical(1);
		print $json->pretty->encode(\@J);
	}
}


sub usage($) {
	my ($err) = @_;

	my $FH = $err ? \*STDERR : \*STDOUT;

	print $FH <<EOH
Usage: $PROGNAME [-_h] [-c char] [-o name] [-r sep] [-s sep]
       -_       replace spaces in keys with '_'
       -c char  treat 'char' as a comment character
       -h       print this help and exit
       -o name  specify the object name
       -r sep   specify the record separator
       -s sep   specify the key-value separator
EOH
;

}

###
### Main
###

init();
parseInput();
printOutput();

exit(0);
